tests/epyccel/test_generic_functions.py

```diff
@@ -105,3 +105,131 @@ def test_multi_tmplt_2(language):
+    assert f1(5) == f1(5)
+    assert f1(5.3, True) == f1(5.3, True)
+    assert f1(5, True) == f2(5, True)
+
+def test_default_var_4(language):
+    f1 = epyccel(mod.default_var_4 , language = language)
+    f2 = mod.default_var_4
+
+    assert f1(5, 5) == f2(5, 5)
+    assert f1(5.3, 5) == f2(5.3, 5)
+    assert f1(4) == f2(4)
+    assert f1(5.2) == f2(5.2)
+
+def test_optional_var_1(language):
+    f1 = epyccel(mod.optional_var_1 , language = language)
+    f2 = mod.optional_var_1
+
+    assert f1(5.3) == f2(5.3)
+    assert f1(5) == f2(5)
+    assert f1(5.3, 2) == f2(5.3, 2)
+    assert f1(5, 2) == f2(5, 2)
+
+def test_optional_var_2(language):
+    f1 = epyccel(mod.optional_var_2 , language = language)
+    f2 = mod.optional_var_2
+
+    assert f1(5.3) == f2(5.3)
+    assert f1(5) == f2(5)
+    assert f1(5.3, complex(1, 5)) == f2(5.3, complex(1, 5))
+    assert f1(5, complex(1, 4)) == f2(5, complex(1, 4))
+
+def test_optional_var_3(language):
+    f1 = epyccel(mod.optional_var_3 , language = language)
+    f2 = mod.optional_var_3
+
+    assert f1(5, 5.5) == f2(5, 5.5)
+    assert f1(5.3, 5.5) == f2(5.3, 5.5)
+    assert f1(4) == f2(4)
+    assert f1(5.2) == f2(5.2)
+
+def test_optional_var_4(language):
+    f1 = epyccel(mod.optional_var_4 , language = language)
+    f2 = mod.optional_var_4
+
+    assert f1(complex(5, 4), 5) == f2(complex(5, 4), 5)
+    assert f1(2.2, 5)  == f2(2.2, 5)
+    assert f1(4.2) == f2(4.2)
+    assert f1(complex(4, 6)) == f2(complex(4, 6))
+
+def test_int_types(language):
+    f1 = epyccel(mod.int_types , language = language, verbose=True)
+    f2 = mod.int_types
+
+    assert f1(10, 5) == f2(10, 5)
+    assert f1(np.int(15) , np.int(10)) == f2(np.int(15) , np.int(10))
+    assert f1(np.int16(5), np.int16(4)) == f2(np.int16(5), np.int16(4))
+    assert f1(np.int8(4), np.int8(7)) == f2(np.int8(4), np.int8(7))
+    assert f1(np.int32(155), np.int32(177)) == f2(np.int32(155), np.int32(177))
+    assert f1(np.int64(166), np.int64(255)) == f2(np.int64(166), np.int64(255))
+
+def test_float_types(language):
+    f1 = epyccel(mod.float_types , language = language, verbose=True)
+    f2 = mod.float_types
+
+    assert f1(10.5, 5.5) == f2(10.5, 5.5)
+    assert f1(np.float(15.5) , np.float(10.5)) == f2(np.float(15.5) , np.float(10.5))
+    assert f1(np.float32(155.2), np.float32(177.1)) == f2(np.float32(155.2), np.float32(177.1))
+    assert f1(np.float64(166.6), np.float64(255.6)) == f2(np.float64(166.6), np.float64(255.6))
+
+def test_complex_types(language):
+    f1 = epyccel(mod.complex_types , language = language, verbose=True)
+    f2 = mod.complex_types
+
+    assert f1(complex(1, 2.2), complex(1, 2.2)) == f2(complex(1, 2.2), complex(1, 2.2))
+    assert f1(np.complex(15.5, 2.0) , np.complex(10.5, 3.4)) == f2(np.complex(15.5, 2.0) , np.complex(10.5, 3.4))
+    np.isclose(f1(np.complex64(15.5 + 2.0j) , np.complex64(10.5 + 3.4j)), f2(np.complex64(15.5 + 2.0j) , np.complex64(10.5 + 3.4j)), rtol = 1e-7, atol = 1e-8)
```

Comment:
There is an `assert` missing here. When you put it back I think you will find that the test fails. I am getting a `ComplexWarning` from this line: `Casting complex values to real discards the imaginary part`. I think it is because you are using the python functions `PyComplex_RealAsDouble` and `PyComplex_ImagAsDouble` to unpack the complex. But here we do not have a `PyComplex` but instead a numpy complex.

Can we use this function for unpacking numpy types?
https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.array.html#c.PyArray_ScalarAsCtype

Does this indicate a more fundamental problem with how we treat types given that we have no way to distinguish between a python and a numpy type? If so do you think this problem is limited to the cwrapper or does it have larger consequences? (e.g. for properties)
```
In [1]: import numpy as np

In [2]: a = np.int64(2)

In [3]: len(dir(a))
Out[3]: 136

In [4]: b = 2

In [5]: len(dir(b))
Out[5]: 70
```

Reply:
Here is the link with a more up to date version of numpy:
https://numpy.org/doc/stable/reference/c-api/array.html?highlight=array%20api#c.PyArray_ScalarAsCtype